.github/workflows/build_and_test.yml

```
@@ -0,0 +1,287 @@
+
+      - name: Setup Node.js
+        uses: actions/setup-node@v4
+        with:
+          node-version: '12.x'
+      - name: Check Node.js version
+        run: node --version
+
+      - name: Checkout Mining Integration Tests Repository
+        uses: actions/checkout@v4
+        with:
+          repository: rsksmart/mining-integration-tests
+          ref: ${{ secrets.MINING_INTEGRATION_TESTS_REF }}
+          token: ${{ secrets.GITHUB_TOKEN }}
+          path: mining-integration-tests
+
+      - name: Install Mining Integration Tests dependencies
+        working-directory: mining-integration-tests
+        run: |
+          npm ci
+
+      - name: Change mining-integration-tests bitcoind url in config.json to localhost
+        working-directory: mining-integration-tests
+        run: |
+          jq '.bitcoind.url = "localhost"' config.json > config.json.tmp && mv config.json.tmp config.json
+
+      - name: Generate BTC blocks
+        working-directory: mining-integration-tests
+        run: |
+          node --unhandled-rejections=strict generateBtcBlocks.js
+
+      - name: Setup Java JDK
+        uses: actions/setup-java@v3
+        with:
+          java-version: '17'
+          distribution: 'temurin'
+
+      - uses: actions/cache@v4
+        name: Restore gradle cache
+        id: cache-gradle
+        with:
+          path: |
+            .gradle/caches
+            gradle/wrapper
+          key: ${{ runner.os }}-gradle-${{ hashFiles('**/*.gradle*', '**/gradle-wrapper.properties') }}
+          restore-keys: |
+            ${{ runner.os }}-gradle-
+
+      - name: Get gradle wrapper and build dependencies
+        run: |
+          if [ ! -f gradle/wrapper/gradle-wrapper.jar ]; then
+            ./configure.sh
+            ./gradlew --no-daemon dependencies
+          fi
+
+      - name: Download build artifacts
+        uses: actions/download-artifact@v4
+        with:
+          name: build-files
+          path: |
+            rskj-core/build
+
+      - name: Start RSKj and Run Tests
+        working-directory: mining-integration-tests
+        run: |
+          version=$(tr -d "'\"" < ../rskj-core/src/main/resources/version.properties \
+            | cut -d = -f 2- | paste -sd - -)
+          echo "Using RskJ version $version at ../rskj-core/build/libs/rskj-core-$version-all.jar"
+          java -Drsk.conf.file=./rsk-integration-test.conf -cp ../rskj-core/build/libs/rskj-core-"$version"-all.jar co.rsk.Start --regtest & rskpid=$!
+          export tries=0
+          export MAX_TRIES=10
+          
+          until nc -z 127.0.0.1 4444 || [ $tries -eq $MAX_TRIES ] || ! kill -0 $rskpid 2>/dev/null; do
+            echo "Waiting for RskJ..."
+            export tries=$((tries+1))
+            sleep 1
+          done
+          
+          if ! kill -0 $rskpid 2>/dev/null; then
+            echo "RskJ process has died"
+            exit 1
+          fi
+          
+          npm test
+          kill -9 $rskpid
```

The `export` aren't needed. Checking for the PID of RskJ doesn't make much sense, as it'll connectivity test will also fail in such case. If max attempts are reached, this will abort the run only if the RskJ died. Otherwise it'll try to run nevertheless. Despite being tests and a throwaway environment, `kill -9` is bad.

For all this reasons, I suggest the following instead:
```suggestion
          
          tries=0
          MAX_TRIES=10
          while [ $tries -lt $MAX_TRIES ]; do
            nc -z 127.0.0.1 && break
            echo "Waiting for RskJ..."
            tries=$((tries+1))
            sleep 1
          done
          
          if [ $tries -eq $MAX_TRIES ]; then
            echo "RskJ unreachable after $MAX_TRIES attempts. Aborting." >&2
            exit 1
          fi
          
          npm test
          kill $rskpid
```

---

the exports were needed for some reason. dunno why but it wouldn't recognise it otherwise. I mean, I can try again...

---

suggestion:

Add this logic to a script inside a folder on the `.github` directory. Then you could call this script here instead of having all this logic inside the .yml file.  All content inside the .github folder is accessible inside the github actions steps.

If we update the script, we don't need to update the .yml, would be an abstraction and easy to validate

---

done @lucasvuotto 

Unless it gets more votes I'll keep it in this file @fmacleal if that's alright?